lib/bridge/PublicitySyncer.js

```diff
@@ -0,0 +1,157 @@
+"use strict";
+var Promise = require("bluebird");
+var log = require("../logging").get("PublicitySyncer");
+
+function PublicitySyncer(ircBridge) {
+    this.ircBridge = ircBridge;
+
+    // Cache the mode of each channel, the visibility of each room and the
+    // known mappings between them. When any of these change, any inconsistencies
+    // should be resolved by keeping the matrix side as private as necessary
+    this._visibilityMap = {
+        mappings: {
+            //room_id: ['server #channel1', 'server channel2',...]
+        },
+        channelIsSecret: {
+            // '$server $channel': true | false
+        },
+        roomVisibilities: {
+            // room_id: "private" | "public"
+        }
+    }
+}
+
+// This is used so that any updates to the visibility map will cause the syncer to
+// reset a timer and begin counting down again to the eventual call to solve any
+// inconsistencies in the visibility map.
+var solveVisibilityTimeoutId = null;
+
+PublicitySyncer.prototype.updateVisibilityMap = function(isMode, key, value) {
+    let hasChanged = false;
+    if (isMode) {
+        if (typeof value !== 'boolean') {
+            throw new Error('+s state must be indicated with a boolean');
+        }
+        if (this._visibilityMap.channelIsSecret[key] !== value) {
+            this._visibilityMap.channelIsSecret[key] = value;
+            hasChanged = true;
+        }
+    }
+    else {
+        if (typeof value !== 'string' || (value !== "private" && value !== "public")) {
+            throw new Error('Room visibility must = "private" | "public"');
+        }
+
+        if (this._visibilityMap.roomVisibilities[key] !== value) {
+            this._visibilityMap.roomVisibilities[key] = value;
+            hasChanged = true;
+        }
+    }
+
+    if (hasChanged) {
+        clearTimeout(solveVisibilityTimeoutId);
+
+        log.info('Solving in 10s');
+        solveVisibilityTimeoutId = setTimeout(() => {
+            this._solveVisibility().catch((err) => {
+                log.error(err.message);
+            });
+        }, 10000);
+        return Promise.resolve();
+    }
+
+    return Promise.resolve();
+}
+
+/* Solve any inconsistencies between the currently known state of channels '+s' modes
+   and rooms 'visibility' states. This does full graph traversal to prevent any +s
+   channels ever escaping into a 'public' room. This function errs on the side of
+   caution by assuming an unknown channel state is '+s'. This just means that if the
+   modes of a channel are not received yet (e.g when no virtual user is in said channel)
+   then the room is assumed secret (+s).
+
+   The bare minimum is done to make sure no private channels are leaked into public
+   matrix channels. If ANY +s channel is somehow being bridged into a room, that room
+   is updated to private. If ALL channels somehow being bridged into a room, that
+   room is allowed to be public.
```

Comment:
Missing a few words? "If ALL channels being bridged into a room are NOT +s, that room is allowed to be public."